code/__HELPERS/weighted_lists.dm

```diff
@@ -0,0 +1,48 @@
+/**
+ * Pick From Weighted Lists
+ * The first value of a list is the weight, which is equal to the number of entries into the pick.
+ * Additional entries are either "end_point"s which are endpoints for spawn selection OR they are unique named lists with their own weight element.
+ * This supports nested, weighted lists, an example can be found in arcade.dm with "arcade_prize_pool"
+ *
+ * The proc adds the list weights together and rolls a number with the combined weight as the cap.
+ * It then goes through the list deducting the weight values of each entry in sequence until it hits 0 or less.
+ * Once the value hits 0 the loop breaks and the list selected provides it's "element_list" as a result.
+ *
+ * Arguments
+ * list to pick from ~ this is the list it will be reading and moving through
+ *
+ * Example List Format
+ *  list/"lootdrop"(
+ *	"common" = list(
+ *		"weight" = 4,
+ *		"materials" = list(
+ *			"weight" = 2,
+ *			"end_point" = (iron, glass, plastic)
+ *			)
+ *		"weapon" = list(
+ *			"weight" = 1,
+ *			"end_point" = (sword, hammer, gun)
+ *			)
+ *		)
+ *	"rare" = list(
+ *		"weight" = 1
+ *		"end_point" = (gold, loot, gems)
+ *		)
+ *  )
+ */
+
```

Comment:
I get why this was done but loot spawners already handle this way more gracefully. Instead of this strange pseudo-struct going on, use lists or typepaths as the keys in assoc lists with the values being the weight.
Examples: https://github.com/tgstation/tgstation/blob/master/code/_globalvars/lists/maintenance_loot.dm

I would say just pull the pick code off /obj/effect/spawner/lootdrop and put it into its own proc then reformat your lists.